### DISCUSSION TITLE
Can we achieve Zero down time when there is setup:upgrade is happening

### DISCUSSION THREAD

**Discussion by ghost** [KiweeEu/magento2-on-kubernetes#292] [Q&A]
is there any way we can achieve  Zero downtime  when we are upgrading the DB ?

**Comment by macieklewkowicz** [comment]
It should be possible, however not with vanilla Magento.

When there's anything to update in the database, `setup:upgrade` and `app:config:import` are performed automatically on Magento Pod initialization (https://github.com/KiweeEu/magento2-on-kubernetes/blob/2.4.6-p4-3/deploy/bases/app/magento-web.yaml#L90-L105). This would be sufficient to allow zero-downtime deployments if not for how Magento handles schema and data updates.

Because by default Magento displays an error on all pages when any schema versions in the database don't match those in the code, the new version won't work before `setup:upgrade`, and the old version breaks as soon as `setup:upgrade` is performed.

This means that you can either:
* have a brief maintenance window when database updates are being performed (as recommended in https://experienceleague.adobe.com/en/docs/commerce-operations/configuration-guide/deployment/technical-details#recommended-pipeline-deployment-workflow), or
* modify Magento's schema version handling so that it allows mismatch between schema and data versions in the database and the code (there are 3rd-party modules that do this but I cannot recommend any particular one)

**Comment by ghost** [comment]
issue resolved
